### ISSUE TITLE
Computing batch-wise instead of sample-wise dCov

### ISSUE THREAD

**Issue by benquick123** [zjsong/CDNet#1]
Hi,

I'm trying to reuse your code for dCov and I noticed that your implementation is diverging from the paper's description. To be more concrete, I believe the distance covariance should be sample wise and not batch wise -- which it is not, since `cent_dist_x` and `cent_dist_y` in the line below are both scalars and not vectors of size = `batch_size`.

https://github.com/zjsong/CDNet/blob/914770230b9629db1d0aac97686c42b26956ee1c/MNIST/src/distance_covariance.py#L35

Would you please explain, why there is a descrepancy and if this is a bug or intentional implementation.

**Comment by zjsong** [comment]
> Hi,
> 
> I'm trying to reuse your code for dCov and I noticed that your implementation is deverging from the paper's description. To be more concrete, I believe the distance covariance should be sample wise and not batch wise -- which it is not, since `cent_dist_x` and `cent_dist_y` in the line below are both scalars and not vectors of size = `batch_size`.
> 
> https://github.com/zjsong/CDNet/blob/914770230b9629db1d0aac97686c42b26956ee1c/MNIST/src/distance_covariance.py#L35
> 
> Would you please explain, why there is a descrepancy and if this is a bug or intentional implementation.

Hi @benquick123 , I'm sorry for late reply. 

As you mentioned, the (sample) distance covariance is usually computed and used in a sample wise way. In our experiments, we empirically found if we treat the mini-batch data (batch_size=_N_) as _N_ sample points (say, _N_=100) to compute dCov^2 accordingly, the entire model would perform well. So in this regard, you may see our implementation as an approximation of the sample distance covariance.

On the other hand, both of `cent_dist_x` and `cent_dist_y` are matrixes rather than scalars, and `cent_dist_x * cent_dist_y` will return a new matrix by the element-wise multiplication. The code is implemented using PyTorch's tensor broadcast trick.

Please feel free to let me know if this makes sense.

**Comment by benquick123** [comment]
Hi,

you are right. I don't know what I was watching, but it definitely wasn't `cent_dist_x` and `cent_dist_y` variables, which are matrices.

On the other hand, the problem with using batch-wise distance covariance is that the variance itself changes dramatically in regards to batch-size. See example below:
```python
    cov_loss_fn_bk = dCov2()
    for batch_size in 2 ** torch.arange(1, 9):
        covs = []
        for it in range(10):
            a = torch.rand((batch_size, 4, 3))
            b = torch.rand((batch_size, 12))
            covs.append(cov_loss_fn_bk(a, b))
        print("batch_size: %d, mean_dist_cov: %f" % (batch_size.numpy(), (sum(covs) / 10).numpy()))
```

Code above yields the following mean distance covariances:
```
batch_size: 2, mean_dist_cov: 0.453576
batch_size: 4, mean_dist_cov: 0.359464
batch_size: 8, mean_dist_cov: 0.216253
batch_size: 16, mean_dist_cov: 0.113293
batch_size: 32, mean_dist_cov: 0.058346
batch_size: 64, mean_dist_cov: 0.029955
batch_size: 128, mean_dist_cov: 0.014944
batch_size: 256, mean_dist_cov: 0.007395
```

As you can see the get smaller and smaller depending on `batch_size`. Although sample-wise distance covariance has its own set of problems (e.g. size of distance matrix is O(n^2)), this is unfortunate since `batch_size` varies the gradient magnitudes and makes a model with multiple losses harder to train and possibly has other undesireable consequences. Did you have to deal with that problem at all?

**Comment by zjsong** [comment]
Hi @benquick123 ,

I think the phenomenon that the value of dCov2 in your example gets smaller and smaller as increasing `batch_size` is not strange. On the contrary, it is reasonable that dCov2 tends to be `0` when `batch_size` takes a bigger value, because `a` and `b` are **uncorrelated** therein. By running your code with some bigger `batch_size`, I obtained the following results:
```
batch_size: 2, mean_dist_cov: 0.423689
batch_size: 4, mean_dist_cov: 0.360079
batch_size: 8, mean_dist_cov: 0.218536
batch_size: 16, mean_dist_cov: 0.112109
batch_size: 32, mean_dist_cov: 0.059277
batch_size: 64, mean_dist_cov: 0.029853
batch_size: 128, mean_dist_cov: 0.014992
batch_size: 256, mean_dist_cov: 0.007480
batch_size: 512, mean_dist_cov: 0.003760
batch_size: 1024, mean_dist_cov: 0.001915
batch_size: 2048, mean_dist_cov: 0.000952
batch_size: 4096, mean_dist_cov: 0.000469
batch_size: 8192, mean_dist_cov: 0.000236
batch_size: 16384, mean_dist_cov: 0.000117
```

Additionally, if we use the same python script to compute the sample dCov2 between two **correlated** variables such as `a` and `a * 2`, the result will finally converge to some value, as empirically shown bellow:
running:
```
    for batch_size in 2 ** torch.arange(1, 15):
        covs = []
        for it in range(10):
            a = torch.randn((batch_size, 4, 3))
            b = a.view(batch_size, -1) * 2
            covs.append(cov_loss_fn_bk(a, b))
        print("batch_size: %d, mean_dist_cov: %f" % (batch_size.numpy(), (sum(covs) / 10).numpy()))
```
outputs:
```
batch_size: 2, mean_dist_cov: 11.046906
batch_size: 4, mean_dist_cov: 9.904627
batch_size: 8, mean_dist_cov: 5.664570
batch_size: 16, mean_dist_cov: 3.545535
batch_size: 32, mean_dist_cov: 2.300043
batch_size: 64, mean_dist_cov: 1.697402
batch_size: 128, mean_dist_cov: 1.379945
batch_size: 256, mean_dist_cov: 1.162790
batch_size: 512, mean_dist_cov: 1.069202
batch_size: 1024, mean_dist_cov: 1.031227
batch_size: 2048, mean_dist_cov: 1.012442
batch_size: 4096, mean_dist_cov: 1.003496
batch_size: 8192, mean_dist_cov: 0.996628
batch_size: 16384, mean_dist_cov: 0.994401
```

In practice, I usually select the `batch_size` value from a given set like `{64, 100, 128, 256, 512, 1024}`. Generally speaking, the bigger the `batch_size`, the more accurate the results, but we should also consider the computational burden. So you know what I mean. :wink:

**Comment by benquick123** [comment]
Ok, great, it makes perfect sense to me now. Thanks.
